### DISCUSSION TITLE
Add a method to check if logger is enabled based on severity?

### DISCUSSION THREAD

**Discussion by ThomsonTan** [open-telemetry/opentelemetry-cpp#1945] [General]
In the current log API in opentelemetry-cpp, like EmitLogRecord below, `LogRecord` is created and passed to processors directly. minSeverity should be checked in the beginning to avoid all the heavy operations like log information calculation and log record creation. I think the best thing is we provide a wrapper function like `logger::Enabled(severity)` as a fast pre-check before producing any log data. .NET has something similar [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/logs/getting-started/Program.cs#L38).

https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/logs/logger.h#L163

**Comment by owent** [comment]
Good point, many log library have functions to set and get log level, though it seems not be a part of specification.

**Comment by lalitb** [comment]
The otel specification doesn't provide any logging API which can be directly called through instrumentation library, as indicated [here](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md) -

``
Note: this document defines a log backend API. The API is not intended to be called by application developers directly. It is provided for logging library authors to build [Appenders](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md#how-to-create-log4j-style-appender), which use this API to bridge between existing logging libraries and the OpenTelemetry log data model.
``

So it ONLY defines `log backend API` and `log SDK`. Most of the languages have their own native logging APIs, and the otel `log backend API` enables these languages SIGs to develop Appenders to connect their `native APIs` to `logging SDK`.

`<native-api>`   --> `<appender using log backend API>` --> `logging SDK`  -> `otel exporters`

The get/set level are generally part of these language provided `native-api`. E.g., for python:
```python
import logging
logging.basicConfig(filename='example.log', encoding='utf-8', level=logging.DEBUG)
```

This is probably the reason why we don't have any get and set log levels as part of this `log backend API`.

In case of C++ language, the application developers may be using an existing logging framework (say) `spdlog` or `log4cxx`. Someone needs to then develop appenders using `log backend API` to emit logs through OpenTelemetry provided exporters:

  `<spdlog/log4cxx>`  --> `<appender using log backend API>`  --> `logging SDK` -> `otel-exporters`

If we want to support configuring log-level as part of otel-cpp, we can provide a similar convenience library/wrapper at api level with basic functions to emit logs, set/get level,  something like:

   `otel::init_log()`  -> configures log provider and logger
    `otel::set_log_level(severity)` - Set log level
   `otel::log()` --> Check log level, create logRecord and use Logger::Emit() 


~~If we plan to provide this convenience library/wrapper, this is something can be done once our `logs backend API` and `log SDK` are stable.~~

I think it would be good idea to provide this convenience library/wrapper once our `logs backend API` and `log SDK` are stable.

**Comment by ssams** [comment]
I've seen that as of https://github.com/open-telemetry/opentelemetry-cpp/pull/2094#issuecomment-1512766919 there is an (internal) `SetMinimumSeverity` on the logger, which is according to the comment supposed to be configurable via the SDK. However, I currently don't see any functionality on the SDK `LoggerProviderFactory` / `LoggerProvider` which would allow to influence this - am I overlooking something or is still missing / in progress?
